### PR TITLE
Fused lerp operation & lerp_tile LLK

### LINKED ISSUE #36763: [4xGLX Wan] CFG update on device
We have a set of operations which occurs once per denoising step:
```python
permuted_noise_pred = permuted_noise_uncond + current_guidance_scale * (
    permuted_noise_pred - permuted_noise_uncond
)
```

[pytorch reference](https://github.com/Wan-Video/Wan2.2/blob/main/wan/text2video.py#L351)
[tt-dit reference](https://github.com/tenstorrent/tt-metal/blob/main/models/tt_dit/pipelines/wan/pipeline_wan.py#L865)

permuted_noise_uncond and permuted_noise_pred: [1, 1, S, D]
current_guidance_scale: scalar
D = 64
S = 75776, fractured across SP_AXIS which is 8 or 32 devices for single or quad GLX. That means this single-chip op will operate on S=9472 or S=2368 when S is fractured.

Inputs are ~~bf16~~ fp32 and the calculation and all intermediates must be full fp32 precision. The output should also be fp32 precision. Please avoid passing intermediates through srcA and srcB, since truncation to TF32 will impact correctness.

### PR DESCRIPTION
**Pull request by nmauriceTT**
### Ticket
#36763

### Problem description
For Wan2.2, we want to perform the following:

```
permuted_noise_pred = permuted_noise_uncond + current_guidance_scale * (
    permuted_noise_pred - permuted_noise_uncond
)
```
This is equivalent to `lerp(permuted_noise_uncond, permuted_noise_pred, current_guidance_scale)`.
However, `ttnn.lerp` is defined as a composite and may be lossy due to `ttnn::add`, `ttnn::subtract` and `ttnn::multiply`, especially with bfloat16 data. 

Fusing the operations should also improve performance.

### What's changed
- Add fused lerp implementation
- Update `ttnn.lerp` to take `output_tensor` 
- Add `lerp_tile` LLK
- Update tests for lerp, tests now use ULP instead of PCC.
- Add shape for Wan2.2 in test.

#### Performance

Performance (Wormhole N150)

On 9472 x 64 tensor in DRAM:

dtype | Branch |  Total Duration [ms]
-- | -- | --
bfloat16 | main | 8813
bfoat16 | new | 1495
float32 | main | 15461
float32 | new | 2761

### Checklist

- [x] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:nmaurice/36763-fused-lerp)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:nmaurice/36763-fused-lerp) (unrelated failure, also in other [branches](https://github.com/tenstorrent/tt-metal/actions/runs/21864535471/job/63102851430))
- [x] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:nmaurice/36763-fused-lerp) (executes `test_lerp.py`)
- [x] New/Existing tests provide coverage for changes

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:nmaurice/36763-fused-lerp)
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:nmaurice/36763-fused-lerp)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:nmaurice/36763-fused-lerp)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

File: tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py [hunk 1]
```diff
@@ -2,17 +2,21 @@
 
 # SPDX-License-Identifier: Apache-2.0
 
+import csv
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py]: Removed unused import

File: tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py [hunk 2]
```diff
@@ -2,17 +2,21 @@
...
+import os
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py]: Removed unused imports

File: tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py [hunk 3]
```diff
@@ -2,17 +2,21 @@
...
+
 import pytest
 
 import torch
 
 import ttnn
 
 from math import pi
-from tests.ttnn.utils_for_testing import assert_with_pcc
+from models.common.utility_functions import comp_ulp
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py]: Removed unused imports.

File: ttnn/cpp/ttnn/operations/eltwise/ternary/ternary.cpp [hunk 4]
```diff
@@ -334,4 +334,69 @@ Tensor AddcdivOperation::invoke(
...
+    log_debug(tt::LogOp, "Lerp LLK - TTS (scalar weight)");
+    return ttnn::prim::ternary(
+        TernaryOpType::LERP,
+        input,
+        end,
+        weight,
+        ternary_utils::determine_output_dtype(output, input.dtype()),
+        ternary_utils::determine_memory_config(memory_config, input.memory_config()),
+        output,
+        std::nullopt);
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/eltwise/ternary/ternary.cpp:365]: Added validation

File: ttnn/cpp/ttnn/operations/eltwise/ternary/ternary.cpp [hunk 5]
```diff
@@ -334,4 +334,69 @@ Tensor AddcdivOperation::invoke(
...
+    bool is_any_input_block_format =
+        is_block_float(input.dtype()) || is_block_float(end.dtype()) || is_block_float(weight.dtype());
+    bool is_subtile_bcast = (broadcast_type == TernaryBroadcastType::ROW_BCAST) ||
+                            (broadcast_type == TernaryBroadcastType::COL_BCAST) ||
+                            (broadcast_type == TernaryBroadcastType::SCALAR_BCAST);
+    bool is_input_int32 =
+        (input.dtype() == DataType::INT32) && (end.dtype() == DataType::INT32) && (weight.dtype() == DataType::INT32);
+
+    if (is_invalid_bcast(broadcast_type) || (is_any_input_block_format && is_subtile_bcast) || is_input_int32) {
+        log_debug(tt::LogOp, "Lerp Fallback - TTT (tensor weight)");
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/eltwise/ternary/ternary.cpp:386]: Added support for preallocated_tensor in composite operation

File: ttnn/cpp/ttnn/operations/eltwise/ternary/ternary_nanobind.cpp [hunk 6]
```diff
@@ -202,6 +202,7 @@ void bind_ternary_lerp(nb::module_& mod, const ternary_operation_t& operation, c
 
         Keyword Args:
             memory_config (ttnn.MemoryConfig, optional): memory configuration for the operation. Defaults to `None`.
+            output_tensor (ttnn.Tensor, optional): preallocated output tensor. Defaults to `None`.
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/eltwise/ternary/ternary_nanobind.cpp:204]: Updated docstring

File: ttnn/cpp/ttnn/operations/eltwise/ternary/ternary.cpp [hunk 7]
```diff
@@ -334,4 +334,69 @@ Tensor AddcdivOperation::invoke(
...
+    bool is_any_input_block_format = is_block_float(input.dtype()) || is_block_float(end.dtype());
+    bool is_subtile_bcast = (broadcast_type == TernaryBroadcastType::ROW_BCAST) ||
+                            (broadcast_type == TernaryBroadcastType::COL_BCAST) ||
+                            (broadcast_type == TernaryBroadcastType::SCALAR_BCAST);
+    bool is_input_int32 = (input.dtype() == DataType::INT32) && (end.dtype() == DataType::INT32);
+
+    if (is_invalid_bcast(broadcast_type) || (is_any_input_block_format && is_subtile_bcast) || is_input_int32) {
+        log_debug(tt::LogOp, "Lerp Fallback - TTS (scalar weight)");
+        return _lerp_overload(input, end, weight, memory_config);
+    }
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/eltwise/ternary/ternary.cpp:354]: Added support for preallocated tensor in both composite fallbacks.

File: tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py [hunk 8]
```diff
@@ -8,69 +8,121 @@
...
 @pytest.mark.parametrize("weight", [0.75])
 def test_lerp_tensor_b(device, h, w, weight):
-    run_lerp_test_tensor(device, h, w, 1, 80, 99, weight, ttnn.lerp, torch.lerp, pcc=0.999)
+    run_lerp_test_tensor(device, h, w, 1, 80, 99, weight, ttnn.lerp, ulp_threshold=2)
+
+
+@pytest.mark.parametrize("h", [64])
+@pytest.mark.parametrize("w", [9472])
+@pytest.mark.parametrize("weight", [0.75])
+def test_lerp_bf16_inputs_fp32_preallocated_output(device, h, w, weight):
```
Comment (cglagovichTT) [inline tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py]: Can we add a test for fp32 inputs, fp32 output?

**Review by cglagovichTT — APPROVED** [review]

File: tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py (hunk 8 above)
Comment (nmauriceTT) [reply tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py]: Fixed: I have added tests cases for float32 inputs.

**Review by fvranicTT — APPROVED** [review]
LLK part looks good. If you need more perf, you'll likely have to write TTI/TT by hand. You could probably use `TTI_SFP_STOCH_RND` for rounding too.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_lerp.h [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+    // size of each tile in Dest is 64/SFP_DESTREG_STRIDE = 32 rows when using sfpi to load/store
+    constexpr uint dst_tile_size_sfpi = 32;
+    // lerp: out = input + weight * (end - input)
+#pragma GCC unroll 8
+    for (int d = 0; d < ITERATIONS; d++) {
+        sfpi::vFloat in0 = sfpi::dst_reg[dst_index_in0 * dst_tile_size_sfpi];
+        sfpi::vFloat in1 = sfpi::dst_reg[dst_index_in1 * dst_tile_size_sfpi];
+        sfpi::vFloat in2 = sfpi::dst_reg[dst_index_in2 * dst_tile_size_sfpi];
+        sfpi::vFloat result = in0 + in2 * (in1 - in0);
```
Comment (nvelickovicTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_lerp.h:31]: Since this is for Wan, performance is probably important.
Just to put here that for BFloat16 this could be implemented on FPU.

**Review by nvelickovicTT — APPROVED** [review]

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_lerp.h (hunk 9 above)
Comment (nmauriceTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_lerp.h:31]: In this specific case, we want full precision for the CFG update (c.f. #36763). That's also why I went with LLK rather than a metal kernel: we can keep DST registers as bf16 without losing accuracy from intermediate quantization errors (initial plan was to use bf16 inputs and bf16 or fp32 outputs). 

For performance: I ran a quick test with an empty LLK on the Wan 2.2 shapes ([9472, 64] tensor in DRAM). The performance 'gains' was small (<3% faster). The op' is likely still memory bound (although, manual TTI optimization might still be useful for L1 tensors).

File: tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py [hunk 10]
```diff
@@ -8,69 +8,132 @@
...
 
-from math import pi
-from tests.ttnn.utils_for_testing import assert_with_pcc
+from tests.ttnn.utils_for_testing import assert_with_ulp
 
 
-def run_lerp_test_float(device, h, w, low, high, end, weight, ttnn_function, torch_function, pcc=0.9999):
-    torch_input_tensor_a = torch.linspace(low, high, steps=h * w, dtype=torch.bfloat16).reshape((h, w))
-    torch_input_tensor_b = torch.full((h, w), end, dtype=torch.bfloat16)
+def run_lerp_test_scalar_weight(
```
Comment (mateusznowakTT) [inline tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py]: These two functions look almost exactly the same. I would squash them and add flag to distinguish scalar from tensor, and put those 3 lines that differ inside if

**Review by mateusznowakTT — CHANGES_REQUESTED** [review]

**Review by KalaivaniMCW — APPROVED** [review]

File: tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py (hunk 10 above)
Comment (nmauriceTT) [reply tests/ttnn/nightly/unit_tests/operations/eltwise/test_lerp.py]: Fixed. I've fused both functions and added a `use_scalar_weight` flag to switch between scalar weight <-> tensor weight.

**Review by dchenTT — APPROVED** [review]

**Review by mateusznowakTT — APPROVED** [review]
